response_test.go

```diff
@@ -829,6 +830,205 @@ func TestMarshalMany_InvalidIntefaceArgument(t *testing.T) {
+			expectedRes: false,
+		},
+		test{
+			scenario:    "not embedded",
+			input:       reflect.StructField{Type: structType},
+			expectedRes: false,
+		},
+	}
+
+	for _, test := range tests {
+		res := isEmbeddedStruct(test.input)
+		if res != test.expectedRes {
+			t.Errorf("Scenario -> %s\nGot -> %v\nExpected -> %v\n", test.scenario, res, test.expectedRes)
+		}
+	}
+}
+
+func TestShouldIgnoreField(t *testing.T) {
+	type test struct {
+		scenario    string
+		input       string
+		expectedRes bool
+	}
+
+	tests := []test{
+		test{
+			scenario:    "opt-out",
+			input:       annotationIgnore,
+			expectedRes: true,
+		},
+		test{
+			scenario:    "no tag",
+			input:       "",
+			expectedRes: false,
+		},
+		test{
+			scenario:    "wrong tag",
+			input:       "wrong,tag",
+			expectedRes: false,
+		},
+	}
+
+	for _, test := range tests {
+		res := shouldIgnoreField(test.input)
+		if res != test.expectedRes {
+			t.Errorf("Scenario -> %s\nGot -> %v\nExpected -> %v\n", test.scenario, res, test.expectedRes)
+		}
+	}
+}
+
+func TestIsValidEmbeddedStruct(t *testing.T) {
+	type foo struct{}
+
+	structType := reflect.TypeOf(foo{})
+	stringType := reflect.TypeOf("")
+	if structType.Kind() != reflect.Struct {
+		t.Fatal("structType.Kind() is not a struct.")
+	}
+	if stringType.Kind() != reflect.String {
+		t.Fatal("stringType.Kind() is not a string.")
+	}
+
+	type test struct {
+		scenario    string
+		input       reflect.StructField
+		expectedRes bool
+	}
+
+	tests := []test{
+		test{
+			scenario:    "success",
+			input:       reflect.StructField{Anonymous: true, Type: structType},
+			expectedRes: true,
+		},
+		test{
+			scenario:    "opt-out",
+			input:       reflect.StructField{Anonymous: true, Tag: "jsonapi:\"-\"", Type: structType},
+			expectedRes: false,
+		},
+		test{
+			scenario:    "wrong type",
+			input:       reflect.StructField{Anonymous: true, Type: stringType},
+			expectedRes: false,
+		},
+		test{
+			scenario:    "not embedded",
+			input:       reflect.StructField{Type: structType},
+			expectedRes: false,
+		},
+	}
+
+	for _, test := range tests {
+		res := (isEmbeddedStruct(test.input) && !shouldIgnoreField(test.input.Tag.Get(annotationJSONAPI)))
+		if res != test.expectedRes {
+			t.Errorf("Scenario -> %s\nGot -> %v\nExpected -> %v\n", test.scenario, res, test.expectedRes)
+		}
+	}
+}
+
+func TestMarshalUnmarshalCompositeStruct(t *testing.T) {
+	type Thing struct {
+		ID   int    `jsonapi:"primary,things"`
+		Fizz string `jsonapi:"attr,fizz"`
+		Buzz int    `jsonapi:"attr,buzz"`
+	}
+
+	type Model struct {
+		Thing
+		Foo string `jsonapi:"attr,foo"`
+		Bar string `jsonapi:"attr,bar"`
+		Bat string `jsonapi:"attr,bat"`
+	}
+
+	model := &Model{}
+	model.ID = 1
+	model.Fizz = "fizzy"
+	model.Buzz = 99
+	model.Foo = "fooey"
+	model.Bar = "barry"
+	model.Bat = "batty"
+
+	buf := bytes.NewBuffer(nil)
+	if err := MarshalOnePayload(buf, model); err != nil {
```

Comment:
Needs to be updated to `MarshalPayload` (if you want `[]byte`) or `Marshal` if you want a `Payloader`